v0/destinations/pinterest_tag/utils.js

```diff
@@ -0,0 +1,300 @@
+/* eslint-disable no-param-reassign */
+const sha256 = require("sha256");
+const { EventType } = require("../../../constants");
+const {
+  CustomError,
+  constructPayload,
+  isDefinedAndNotNull,
+  isDefined,
+  getHashFromArray
+} = require("../../util");
+const { COMMON_CONFIGS } = require("./config");
+
+const VALID_ACTION_SOURCES = ["app_android", "app_ios", "web", "offline"];
+
+const ecomEventMaps = [
+  {
+    src: ["order completed"],
+    dest: "Checkout"
+  },
+  {
+    src: ["product added"],
+    dest: "AddToCart"
+  },
+  {
+    src: ["products searched", "product list filtered"],
+    dest: "Search"
+  }
+];
+
+const USER_NON_ARRAY_PROPERTIES = ["client_user_agent", "client_ip_address"];
+
+/**
+ *
+ * @param {*} userPayload Payload mapped from user fields
+ * @returns
+ * Further Processing the user fields following the instructions of Pinterest Conversion API
+ * Ref: https://s.pinimg.com/ct/docs/conversions_api/dist/v3.html
+ */
+const processUserPayload = userPayload => {
+  let formatValue = "";
+  Object.keys(userPayload).forEach(key => {
+    switch (key) {
+      case "em":
+        formatValue = userPayload[key].toString().toLowerCase();
+        userPayload[key] = [sha256(formatValue)];
+        break;
+      case "ph":
+      case "zp":
+        // zip fields should only contain digits
+        formatValue = userPayload[key].toString().replace(/[^0-9]/g, "");
+        if (key === "ph") {
+          // phone numbers should not contain leading zeros
+          formatValue = formatValue.replace(/^0+/, "");
+        }
+        userPayload[key] = [sha256(formatValue)];
+        break;
+      case "ct":
+      case "st":
+      case "country":
+      case "ge":
+      case "db":
+      case "ln":
+      case "fn":
+      case "hashed_maids":
+        userPayload[key] = [sha256(userPayload[key])];
+        break;
+      default:
+        userPayload[key] = String(userPayload[key]);
+    }
+  });
+  return userPayload;
+};
+
+/**
+ *
+ * @param {*} message
+ * @returns opt_out status
+ *
+ */
+
+const deduceOptOutStatus = message => {
+  const adTrackingEnabled = message.context?.device?.adTrackingEnabled;
+  const attStatus = message.context?.device?.attStatus;
+  let optOut;
+
+  // for android
+  if (isDefinedAndNotNull(adTrackingEnabled)) {
+    if (adTrackingEnabled === true) {
+      optOut = false;
+    } else if (adTrackingEnabled === false) {
+      optOut = true;
+    }
+  }
+
+  // for ios
+  if (isDefinedAndNotNull(attStatus)) {
+    optOut = attStatus === 3 ? true : false;
```

Comment:
optOut = attStatus === 3;

Reply:
deleted this logic, as discussed, the field name is same for ios and android sdks.